rse-py-package.Rmd

```diff
@@ -0,0 +1,1460 @@
+out into its own tool, `bin/check_zipf.py`:
+
+<!-- src="rse-py-package/08/bin/check_zipf.py" -->
+```python
+#!/usr/bin/env python
+import sys
+import zipfpy.check
+
+USAGE = '''zipf num [num...]: are the given values Zipfy?'''
+
+if __name__ == '__main__':
+    if len(sys.argv) == 1:
+        print(USAGE)
+    else:
+        values = [int(a) for a in sys.argv[1:]]
+        result = zipfpy.check.is_zipf(values)
+        print('{}: {}'.format(result, values))
+    sys.exit(0)
+```
+
+That means that the `zipfpy/check.py` file now contains only
+the routines `is_zipf()` and `is_unsorted_zipf()`:
+
+<!-- src="rse-py-package/08/zipfpy/check.py" -->
+```python
+from pytest import approx
+from .generate import make_zipf
+from . import RELATIVE_ERROR
+
+
+def is_zipf(hist, rel=RELATIVE_ERROR):
+    assert len(hist) > 0, 'Cannot test Zipfiness without data'
+    scaled = [h/hist[0] for h in hist]
+    perfect = make_zipf(len(hist))
+    return scaled == approx(perfect, rel=rel)
+
+
+def is_unsorted_zipf(hist, rel=RELATIVE_ERROR):
+    sortedhist = sorted(hist)
+    return is_zipf(sortedhist)
+```
+
+Now we make sure that setuptools knows about this script and
+will install it:
+
+<!-- src="rse-py-package/08/setup.py" -->
+```python
+from setuptools import setup
+
+setup(
+    name='zipfpy',
+    version='0.1',
+    author='Greg Wilson',
+    packages=['zipfpy'],
+    scripts=['bin/check_zipf.py']
+)
+```
+
+
+<!-- used="rse-py-package/08/requirements.txt" -->
+<!-- used="rse-py-package/08/test_zipfpy.py" -->
+<!-- used="rse-py-package/08/zipfpy/__init__.py" -->
+<!-- used="rse-py-package/08/zipfpy/check.py" -->
+<!-- used="rse-py-package/08/zipfpy/generate.py" -->
+
+Now we we'll install the updated package in a virtual environment:
+
+```text
+coredump-2:08 ljdursi$ virtualenv test_script
+..
+Installing setuptools, pip, wheel...
+done.
+$ source test_script/bin/activate
+(test_script) $ pip install -r requirements.txt
+...
+Installing collected packages: atomicwrites, attrs, zipp, importlib-metadata, more-itertools, six, pyparsing, packaging, pluggy, py, wcwidth, pytest
+pip Successfully installed atomicwrites-1.3.0 attrs-19.1.0 importlib-metadata-0.18 more-itertools-7.0.0 packaging-19.0 pluggy-0.12.0 py-1.8.0 pyparsing-2.4.0 pytest-4.6.3 six-1.12.0 wcwidth-0.1.7 zipp-0.5.1
+ins
+(test_script) $ pip install .
+...
+Building wheels for collected packages: zipfpy
+  Building wheel for zipfpy (setup.py) ... done
+  Stored in directory: /private/var/folders/44/k2tfp8d12h7ggxc0xz8phfmr0000gn/T/pip-ephem-wheel-cache-0z_4yd1b/wheels/1c/84/f1/3dedca5ae7f979ead1ca046473b4a1f13552e873ccf1c1d958
+Successfully built zipfpy
+Installing collected packages: zipfpy
+Successfully installed zipfpy-0.1
+
+(test_script) $ check_zipf.py 100 50 33 25
+True: [100, 50, 33, 25]
+```
+
+FIXME: use consistent environment directories in the above examples
+
+FIXME: clever exercises
+
+## How should I document my package for others? {#rse-py-package-document}
```

Comment:
I think the material on documenting software belongs in a lesson of its own - what do you think?

Reply:
Yup